### PR TITLE
CB-6267 Add AWS Spot parameters to Environment creation request

### PR DESCRIPTION
**Pull request by Bajzathd**

### CODE REVIEW

File: environment-api/src/main/java/com/sequenceiq/environment/api/doc/environment/EnvironmentModelDescription.java [hunk 1]
```diff
@@ -72,6 +72,10 @@
     public static final String PARENT_ENVIRONMENT_NAME = "Parent environment name";
     public static final String PARENT_ENVIRONMENT_CLOUD_PLATFORM = "Parent environment cloud platform";
 
+    public static final String FREEIPA_AWS_PARAMETERS = "Aws specific FreeIpa parameters";
+    public static final String FREEIPA_AWS_SPOT_PARAMETERS = "Aws spot instance related parameters.";
+    public static final String FREEIPA_AWS_SPOT_PERCENTAGE = "Percentage of spot instances launched in freeipa instance group";
```
Comment (doktoric) [inline environment-api/src/main/java/com/sequenceiq/environment/api/doc/environment/EnvironmentModelDescription.java:77]: FreeIpa

File: environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/response/FreeIpaResponse.java [hunk 2]
```diff
@@ -11,6 +11,9 @@
     @ApiModelProperty(EnvironmentModelDescription.FREEIPA_INSTANCE_COUNT_BY_GROUP)
     private Integer instanceCountByGroup = 1;
 
+    @ApiModelProperty(EnvironmentModelDescription.FREEIPA_AWS_SPOT_PERCENTAGE)
+    private Integer spotPercentage;
```
Comment (doktoric) [inline environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/response/FreeIpaResponse.java:15]: can we use the same structure here ?

Comment (doktoric) [reply environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/response/FreeIpaResponse.java:15]: like aws.spot.percentage?

File: environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/request/aws/AttachedFreeIpaRequestAwsSpotParameters.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import io.swagger.annotations.ApiModelProperty;
+
+@ApiModel(value = "AttachedFreeIpaRequestAwsSpotParameters")
+public class AttachedFreeIpaRequestAwsSpotParameters {
+
+    @ApiModelProperty(EnvironmentModelDescription.FREEIPA_AWS_SPOT_PERCENTAGE)
+    @Min(value = 0, message = "Spot percentage must be between 0 and 100 percent")
+    @Max(value = 100, message = "Spot percentage must be between 0 and 100 percent")
+    @Digits(fraction = 0, integer = 3, message = "Spot percentage has to be a number")
+    private Integer percentage;
```
Comment (bergerdenes) [inline environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/request/aws/AttachedFreeIpaRequestAwsSpotParameters.java:19]: I used plain integer, not the Integer class

**Comment by doktoric** [comment]
I think a validation would be great for the case when user wants to deploy a single node freeipa because in that case we should only enable 0 or 100

**Comment by Bajzathd** [comment]
> I think a validation would be great for the case when user wants to deploy a single node freeipa because in that case we should only enable 0 or 100

Added the validation

File: environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/request/AttachedFreeIpaRequest.java [hunk 4]
```diff
@@ -17,6 +19,10 @@
     @ApiModelProperty(value = EnvironmentModelDescription.FREEIPA_INSTANCE_COUNT_BY_GROUP)
     private Integer instanceCountByGroup;
 
+    @Valid
```
Comment (bergerdenes) [inline environment-api/src/main/java/com/sequenceiq/environment/api/v1/environment/model/request/AttachedFreeIpaRequest.java:22]: please add @Valid to `private AttachedFreeIpaRequest freeIpa;` line 72 the EnvironmentRequest class

File: environment/src/main/java/com/sequenceiq/environment/environment/dto/FreeIpaCreationDto.java [hunk 5]
```diff
@@ -25,11 +27,16 @@ public int getInstanceCountByGroup() {
...
+    public FreeIpaCreationAwsParametersDto getAws() {
+        return aws;
+    }
+
     @Override
     public String toString() {
         return "FreeIpaCreationDto{" +
             "create='" + create + '\'' +
             "instanceCountByGroup='" + instanceCountByGroup + '\'' +
+            "aws='" + aws + '\'' +
```
Comment (bergerdenes) [inline environment/src/main/java/com/sequenceiq/environment/environment/dto/FreeIpaCreationDto.java:39]: FreeIpaCreationAwsSpotParametersDto does not define toString()

File: environment/src/main/java/com/sequenceiq/environment/parameters/dao/domain/AwsParameters.java [hunk 6]
```diff
@@ -17,6 +17,9 @@
     @Enumerated(EnumType.STRING)
     private S3GuardTableCreation s3guardTableCreation;
 
+    @Column(name = "free_ipa_spot_percentage")
```
Comment (bergerdenes) [inline environment/src/main/java/com/sequenceiq/environment/parameters/dao/domain/AwsParameters.java:20]: minor:  freeipa in one word

File: environment/src/test/java/com/sequenceiq/environment/environment/validation/EnvironmentValidatorServiceTest.java [hunk 7]
```diff
@@ -324,4 +331,30 @@ private Environment aValidEnvirontmentWithParent() {
...
+        ValidationResult validationResult = underTest.validateFreeIpaCreation(freeIpaCreationDto);
+        assertEquals(!valid, validationResult.hasError());
+    }
+
+    private static Stream<Arguments> freeIpaCreationArguments() {
+        return Stream.of(
+                Arguments.of(1, 0, true),
+                Arguments.of(1, 100, true),
+                Arguments.of(1, 50, false),
+                Arguments.of(2, 50, true)
```
Comment (doktoric) [inline environment/src/test/java/com/sequenceiq/environment/environment/validation/EnvironmentValidatorServiceTest.java:356]: can you please add 2 more test ? 
Arguments.of(2, 100, true)
Arguments.of(2, 0, true)

**Pull request merged**